scipy/optimize/_spectral.py

```diff
@@ -0,0 +1,209 @@
+                  fnorm=None, callback=None, disp=False, M=10, eta_strategy=None,
+                  sigma_eps=1e-10, sigma_0=1.0, **unknown_options):
+    r"""
+    Solve nonlinear equation with the DF-SANE method
+
+    Parameters
+    ----------
+    func : callable
+        Function whose root to find
+    x0 : ndarray
+        Initial guess
+    args : tuple
+        Extra parameters to `func`
+    jac
+        Unused, this is a derivative-free solver
+    ftol : float, optional
+        Relative norm tolerance.
+    fatol : float, optional
+        Absolute norm tolerance.
+        Algorithm terminates when ``||func(x)|| < fatol + ftol ||func(x_0)||``.
+    fnorm : callable, optional
+        Norm to use in the convergence check. If None, 2-norm is used.
+    callback : callable, optional
+        Callback function to call on each iteration. Called as
+        ``callable(x, f)`` where ``x`` is the current iterate and
+        ``f`` the residual.
+    maxfev : int, optional
+        Maximum number of function evaluations.
+    disp : bool, optional
+        Whether to print convergence process to stdout.
+    eta_strategy : callable, optional
+        Choice of the ``eta_k`` parameter, which gives slack for growth
+        of ``||F||_2**2``.  Called as ``eta_k = eta_strategy(k, x, F)`` with
+        `k` the iteration number, `x` the current iterate and `F` the current
+        residual. Should satisfy ``eta_k > 0`` and ``sum(eta, k=0..inf) < inf``.
+        Default: ``||F_0||_2**2 / (1 + k)**2`` where F_0 is the initial residual.
+    sigma_eps : float, optional
+        The spectral coefficient is constrained to ``sigma_eps < sigma < 1/sigma_eps``.
+        Default: 1e-10
+    sigma_0 : float, optional
+        Initial spectral coefficient.
+        Default: 1.0
+    M : int, optional
+        Number of iterates to include in the nonmonotonic line search.
+        Default: 10
+
+    References
+    ----------
+    .. [1] "Spectral residual method without gradient information for solving
+           large-scale nonlinear systems of equations." W. La Cruz,
+           J.M. Martinez, M. Raydan. Math. Comp. **75**, 1429 (2006).
+    .. [2] W. La Cruz, Opt. Meth. Software, 29, 24 (2014).
+
+    """
+    _check_unknown_options(unknown_options)
+    if jac is not None:
+        warnings.warn("DF-SANE solver does not use a Jacobian", category=OptimizeWarning)
+
+    nexp = 2
+
+    if eta_strategy is None:
+        # Different choice from [1], as their eta is not invariant
+        # vs. scaling of F.
+        def eta_strategy(k, x, F):
+            # Obtain squared 2-norm of the initial residual from the outer scope
+            return f_0 / (1 + k)**2
+
+    func, x0, x0_shape, F_k, is_complex = _wrap_func(func, x0, args)
+
+    def f(x):
+        if nfev[0] >= maxfev:
+            raise _NoConvergence()
+
+        F = func(x)
+        f = np.linalg.norm(F)**nexp
+        nfev[0] += 1
+        return f, F
```

Comment:
Could the two wrappings `_wrap_func` and `f` here be combined into a single one? I'll make reading easier IMHO.